Enabling showon to work in list options in JForm

GeraintEdwards opened the pull request:
Pull Request for Issue # None.

### Summary of Changes
The showon attribute currently works for fields in Jforms to allow their display to be conditional on the value of another field.  This PR makes some simple changes to enable this functionality to work on list option value too - in other works the display of a specific list option can now be dependent on the value of another field.

The changes required to make this work are minimal:
1.  In libraries/joomla/form/fields/list.php we add pass new attribute name as an option 

`$listoptions['option.attr'] = 'optionattr';`

this means that the list renderer will use the option.attr value we pass in to add additional attributes to the option.

2. In the getOptions method of  libraries/joomla/form/fields/list.php  we look for a 'showon' attribute and if its specified pass it as 'optionattr'

3.  In media/jui/js/cms-uncompressed.js/media/jui/js/cms.js we have to disable the normal animation effect for option values (this does work in Chrome but not in other browsers) and add some special handling to toggle the display of the option and trigger an update in the chosen display where needed).

So all in all minimal changes 

### Testing Instructions

Best test is to edit an existing form e.g. administrator/components/com_content/config.xml and look for the declaration of targetb and add a showon attribute e.g.

`<option value="1" showon="targeta:1">JBROWSERTARGET_NEW</option>`

This make the browser new option in the URL b target drop down dependent on the value of URL a target drop down. 

### Expected result

When you edit the articles options and  set URL A target drop down equal to 'new browser window' this option is available for URL B but otherwise its hidden.

![optionvisible](https://user-images.githubusercontent.com/839810/28498401-913ff60a-6f94-11e7-9761-75e02a20950d.png)

![optionhiden](https://user-images.githubusercontent.com/839810/28498402-91496bc2-6f94-11e7-8253-9f38db7542b0.png)

### Actual result

As explained above.

### Documentation Changes Required

GeraintEdwards committed:
Specify option attribute value for select lists to allow list options to respect showon settings

GeraintEdwards committed:
Javascript changes required to support showon in list options

GeraintEdwards committed:
Javascript changes required to support showon in list options - in firefox too

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
Revert "Fixes codestyling in cms-uncompressed.js (#16947)"

This reverts commit 4f94056f8908d897aea5cad0f2731aeb7110ad1a.

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
JS page formatting

GeraintEdwards committed:
Fussy formatting

tonypartridge commented:
I have tested this item :white_check_mark: successfully on 632ca6cf1486d9015bfac77f7cbd977b1e51a962<br /><br />Working very well. Tested on a few components with different values. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17212">issues.joomla.org/tracker/joomla-cms/17212</a>.</sub>

tonypartridge commented:
@C-Lodder can you review the JS?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/17212">issues.joomla.org/tracker/joomla-cms/17212</a>.</sub>

C-Lodder commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -140,8 +139,17 @@ if (typeof(Joomla) === 'undefined') {
+			if (animate && !target.is('option')) {
 				(showfield) ? target.slideDown() : target.slideUp();
+			} else if (target.is('option')) {
+				target.toggle(showfield);
```
`toggle()` was removed in jQuery 1.9. Please use an alternative such as `toggleClass()` and then use CSS to perform any actions

GeraintEdwards replied:
This isn't new - this is part of the existing code and checking the jquery.js in media/jui/js/jquery.js it is defined at line 7538.

Looking at the jQuery docs I don't see that it is deprecated (http://api.jquery.com/toggle/) - where do you see that it is deprecated?

C-Lodder replied:
![jquery](https://user-images.githubusercontent.com/2019801/28568942-867c55c8-712f-11e7-9d21-5b5293f4cee0.png)

C-Lodder replied:
Even if this wasn't the case, it's always better to use CSS to hide/show elements and just use JS to add/remove a class. Just a faster solution ;)

GeraintEdwards replied:
See the Note in yellow 

> Note: This method signature was deprecated in jQuery 1.8 and removed in jQuery 1.9. jQuery also provides an animation method named .toggle() that toggles the visibility of elements. Whether the animation or the event method is fired depends on the set of arguments passed.

Its the specific signature that is deprecated - the animated method .toggle is not deprecated from what I can see and it takes a boolean argument http://api.jquery.com/toggle/#toggle-display  Running the code through the debugger the jquery-migrate.js script flows through to the underlying animated toggle method in jquery.js v1.12.4 in the core.

C-Lodder replied:
Comment above still applies

GeraintEdwards replied:
The CSS/JS code required to replicate the functionality of the toggle method can be complex -  we can use css('display', none') or the .style equivalent to hide the element but when we reveal we need to know what CSS that was set the relevant element before it was hidden. For some it would be block, others inline and it may even have been set by the template or component CSS file.  See the ShowHide method of jquery.

We could use jquery .show and .hide method but we'd basically be replicating what jquery does already so I'm not sure of the benefits?

linkedoptions is initially called with animate undefined so this change would not only affect list options.


See the ShowHide method in jquery - given how much jquery has been tested I would suggest we are safer

C-Lodder replied:
If it's the animation that's really important here, then why not just `slideToggle()`?

GeraintEdwards replied:
Sorry - I didn't explain very clearly.  We need a non-animated method.

The default method already in the file 

```javascript
(showfield) ? target.slideDown() : target.slideUp();
```

only works in chrome, it doesn't work in Firefox and probably not MSIE so I fell back to the un-animated method which was already in the file 

```javascript
target.toggle(showfield);
```

slideToggle is a shortcut for toggle from what I can see in jquery.js (though the docs don't refer to this).

We could us this code
```javascript
			// If conditions are satisfied show the target field(s), else hide
			// note that animations don't work on list options other than in Chrome
			if (animate && !target.is('option')) {
				(showfield) ? target.slideDown() : target.slideUp();
			} else {
				showfield ? target.show() : target.hide();
				if (target.is('option')) {
					target.toggle(showfield);
					target.attr('disabled', showfield ? false : true);
					// if chosen active for the target select list then update it
					var parent = target.parent();
					if ($('#' + parent.attr('id') + '_chzn').length) {
						parent.trigger("liszt:updated");
						parent.trigger("chosen:updated");
					}
				}
			}
```
 but its still using jQuery.

If the overall goal is to move to native JS away from jQuery then the code needs a wholesale rewrite and a lot of tests added in

tonypartridge replied:
@C-Lodder I agree with @GeraintEdwards on this one, it's supported by the J3.x series given jQuery is loaded and not yet deprecated. J4 can remove and rewrite all of this if need be, but for J3.8 I see no reason it cannot go in as is?

C-Lodder replied:
Meh leave it as is then. Will rewrite for J4

GeraintEdwards committed:
Nominal change to force Jenkins to restart 

Jenkins has failed to run so this commit is only in order to force it to restart

C-Lodder commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -140,8 +139,17 @@ if (typeof(Joomla) === 'undefined') {
+				target.toggle(showfield);
+				target.attr('disabled', showfield ? false : true);
+				// if chosen active for the target select list then update it
+				if ($('#' + target.parent().attr('id') + '_chzn').length) {
```
You're calling `target.parent()` 3 times, so please consider using a variable:

```
var parent = target.parent();
```

GeraintEdwards replied:
Is there a problem with chaining?  Is there preferred javascript coding style that we should be adopting?  I've seen the code formatting standards but nothing about coding style.

Thanks

C-Lodder replied:
I'm not talking about chaining here. Simply that you should do:

```
var parent = target.parent();
if ($('#' + parent.attr('id') + '_chzn').length) {
	parent.trigger("liszt:updated");
	parent.trigger("chosen:updated");
}
```
rather than:
```
if ($('#' + target.parent().attr('id') + '_chzn').length) {
	target.parent().trigger("liszt:updated");
	target.parent().trigger("chosen:updated");
}
```

Just a small performance enhancement.

C-Lodder commented on media/jui/js/cms-uncompressed.js:
```diff
@@ -4,7 +4,7 @@
 
 // Only define the Joomla namespace if not defined.
-if (typeof(Joomla) === 'undefined') {
+if (typeof (Joomla) === 'undefined') {
```
Replace this `if` statement with just:

```
Joomla = window.Joomla || {};
```

GeraintEdwards replied:
no problems

GeraintEdwards committed:
remove duplicated call to target.parent()

GeraintEdwards committed:
Merge origin/showon-listoptions into showon-listoptions

dgrammatiko commented:
@C-Lodder we need to deprecate the cms.js in j3 and remove it in j4. The code in that file should be either a custom element or some plain js **attached** to the relative field, same approach as all the rest fields that have some sort of js functionality (calendar, colour, etc)

GeraintEdwards commented:
The showon elements of the cms.js file relates to an attribute that can be applied to most fields so isn't field specific so if JForm specific as opposed to any specific JField type. 

The list option specific aspect could be extracted to a separate file but we would need a javascript 'architecture' that allowed for this using callback or pseudo class methods ???

I know this isn't the place to have the debate but if we are not going to use jQuery then we will need our own slimmed down javascript library that is unit tested to avoid duplication of coding effort and code that fails in specific scenarios.

GeraintEdwards committed:
code clean up and use show/hide instead of toggle.

dgrammatiko commented:
> will need our own slimmed down javascript library

No. We have ES5 ES6 and web components. No need for anything else!

GeraintEdwards committed:
More haste less speed - had hide/show and toggle but hide/show doesn't work for list options so must stick to toggle

GeraintEdwards commented:
@dgt41 
I presume you mean we would use code like 
```javascript
xpathresults = document.evaluate('//*[@name="' + fieldName + '"]|//*[@name="' + fieldName + '[]"]', document, null, XPathResult.UNORDERED_NODE_SNAPSHOT_TYPE, null)				
```
in J4.0 instead of 
```javascript
$fields   = $('[name="' + fieldName + '"], [name="' + fieldName + '[]"]');
```
That sort of change is quite straightforward for J4.0 and doesn't need a javascript library like jQuery but what about little methods that are used repeatedly in different contexts such as the toggle method discussed above which keeps track of the original display style value, or little animations that would apply to ?  Are we not going to have this type of re-usable utility function in J4.0?  Would these be added to HTMLElement in a namespace safe way (sorry not sure of the jargon here)?

dgrammatiko commented:
@GeraintEdwards modern UI should be done through web component or custom elements. If we do it correctly jQuery will be useless (in fact it's already), because frankly we shouldn't do any animations with js but rather pass that to css (60FPS, 16ms Budget and almost all the current recommendations fail with jQuery, especially if you do the test in a cheap android device). 
Anyways what you're doing here is fine, but for J4 we might have to rethink the whole approach to these fields.

If you're interested about the custom elements, check https://github.com/joomla-projects/custom-elements
and the relative PR for J4: https://github.com/joomla/joomla-cms/pull/17024

GeraintEdwards committed:
cleaning up branch locally

GeraintEdwards committed:
Merge origin/showon-listoptions into showon-listoptions

C-Lodder commented:
Can you sort out the PR so that the only the relevant commits are added?

GeraintEdwards commented:
I think I merged and pushed the wrong branch and now I'm trying to figure it out  how to reverse this

GeraintEdwards committed:
Revert "Merge origin/showon-listoptions into showon-listoptions"

This reverts commit 5aa677ad1b1ec9de3f8b624d769330ea8230beb2, reversing
changes made to 223ef41ce71b30b00c3117839637349c51ca43df.

GeraintEdwards committed:
bring compressed cms.js up to date

GeraintEdwards commented:
Sorted out the bad merge - now the compressed js is up to date with staging and the PR.

What do we need to do to move this forward?

The showon code needs a lot of work:

1. It really slow using jQuery and a large form (I have a prototype version implemented using native query selectors with a pseudo polyfill that falls back to jQuery where the query selector methods don't exist)
2. the toggle needs to be replaced with a CSS transition (this is a very small performance benefit compared to the gains available from improving the finding of conditional elements and processing them on large forms)
3. the toggle/its replacement needs to respect the original CSS state - e.g. a div shouldn't be reset to block if it was an inline-block beforehand
4. the xml syntax should allow for and/or to be combined intelligently with appropriate associative ordering/parenthesis support

A lot of this could be done before 4.0 and I propose putting a PR together to deal with it in the next couple of weeks.

In the meantime cane we move forward with this PR please?

Quy commented on libraries/joomla/form/fields/list.php:
```diff
@@ -144,7 +153,15 @@ protected function getOptions()
 			$tmp['onclick']  = (string) $option['onclick'];
 			$tmp['onchange'] = (string) $option['onchange'];
-
+			
```
Remove tabs

GeraintEdwards replied:
done

GeraintEdwards committed:
remove tabbed blank line

GeraintEdwards committed:
Merge commit '162abb53ec2431465ea97b73ffcbaf6dfe8f3423' into showon-listoptions

Conflicts:
	components/com_content/models/category.php
	components/com_content/views/category/tmpl/blog.xml

GeraintEdwards commented:
Sorry guys - same problem again with the merge commit.  Fixing it now

GeraintEdwards committed:
Revert "Merge commit '162abb53ec2431465ea97b73ffcbaf6dfe8f3423' into showon-listoptions"

This reverts commit 3cb393ce993511edd57c981b5e1a2b70c6f176a6, reversing
changes made to 4b1f22b305291c695e969217a68e459cd6183831.

GeraintEdwards commented:
I'm closing this PR and opening a new one based on a new 'clean' branch.  

I'll include references to this one.

C-Lodder commented:
Before you submit a PR, I suggest you always update your fork so that you're in sync with the main repo.

**Git:**
```
$ git remote add upstream https://github.com/joomla/joomla-cms.git
$ git fetch upstream staging
$ git merge upstream/staging
$ git push origin staging

$ git checkout -b NEW_BRANCH_NAME
// Then do your code changes
```

**Git for Windows/Mac:**
- No idea

GeraintEdwards commented:
Thanks - I was trying to merge upstream using my IDE.  I'll do it on the command line next time !!

PR is now resurrected as https://github.com/joomla/joomla-cms/pull/18998
